server.py

```diff
@@ -0,0 +1,122 @@
+        self.clients_connected = []
+        self.usernames_connected = []
+
+    FORMAT = "utf-8"
+    DATE_FORM = "%d/%m/%Y %H:%M:%S"
+    WELCOME_MSG = "Welcome to the server. Please, type '/help' to get a list of available commands"
+    HELP_MSG = """List of available commands:
+                  /register - to register your username and set password
+                  /login -  to log in to server if you already 
+                            have a registered username
+                  /users - to get a list of connected users
+                  /exit - to disconnect from the server
+                  @<username>: - to send a message to user with a particular 
+                                 username (note: colon is required)
+                  if you are logged in type anything to command prompt to 
+                  broadcast your message to the server and every connected user"""
+
+    def start_server(self):
+        self.server.bind((self.address, self.port))
+        self.server.listen()
+        print("Server started")
+        while True:
+            client, address = self.server.accept()
+            logger.info(f"{address} connected")
+            thread = threading.Thread(target=self.handle_client, args=(client, address))
+            thread.start()
+            print(f"Active connections: {threading.activeCount() - 1}")
+
+    def handle_client(self, client, address):
+        print(f"{address} connected")
+        client.send(self.WELCOME_MSG.encode(self.FORMAT))
+        while True:
+            msg = client.recv(2048).decode(self.FORMAT)
+            if msg == "/help":
+                client.send(self.HELP_MSG.encode(self.FORMAT))
+            elif msg == "/register":
+                self.register(client, self.db_path)
+            elif msg == "/login":
+                self.login(client, self.db_path)
+            elif msg == "/exit":
+                client.close()
+                print(f"{address} disconnected")
+                break
+            elif msg.startswith("@"):
+                self.send_to_client(msg)
+            else:
+                try:
+                    print(f"[{datetime.now().strftime(self.DATE_FORM)}]"
+                          f" {self.usernames_connected[self.clients_connected.index(client)]}:-> {msg}")
+                    self.broadcast(f"[{datetime.now().strftime(self.DATE_FORM)}]"
+                                   f" {self.usernames_connected[self.clients_connected.index(client)]}:->"
+                                   f" {msg}".encode(self.FORMAT))
+                    logger.info(f"{self.usernames_connected[self.clients_connected.index(client)]}"
+                                f" sent a broadcast message")
+                except:
+                    client.send("Only registered users can post. If you are registered try /login".encode(self.FORMAT))
+
+    def register(self, client, db_path):
+        db = Database(db_path)
```

Comment:
why db is not set as instance attribute?

Reply:
Fixed. Users database initializes as an attribute of a SocketServer class now.